### ISSUE TITLE
Data views: Improve the selection column semantics

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#59180]
### Description

HTML tables are meant to contain _data_.
Specifically:
- TD elements are meant to contain data values. Actually, the tag name `TD` is an abbreviation for Table Data.
- Table headers (TH elements) are meant to provide context for the column or row content, establishing a logical relationship between the header text and the table data TDs in that column or row.
- Assistive technology use this semantics. Specifically, screen readers announce the table header text typically eaxh time users enter a column or enter a row.

Historically, WordPress used these table TDs for things that aren't properly 'data types', but that's a separate issue.

To provide this information in a way that is semantically valid and meaningful for users, the table headers in the current implementation of the data views need some improvement, starting with the 'Select / Deselect All' table header.

I do realize the data views are inspired to the list tables used in core so I guess this was basically copied from core. However, what is in core isn't always the most semantic and accessible pattern. Actually, some of the things in core can't be touched because of technical debt or backward compatibility issues. However, the data views is new code that lands in WordPress. As such, we should do our best to provide a better implementation.

- The 'Select / Deselect All' table header contains an interactive control: a checkbox. This is less than ideal and unexpected. Ideally, table headers should only contain the name of the column or row, ie. the 'what" data type the column or row is supposed to contain.
- The checkbox is associated with a label that is the only textual content of the TH. As such, the whole column is named 'Select All' or 'Deselect All' depending on the selection state. This isn't an appropriate name for the column data type.
- As mentioned above, screen reader will announce 'Select All' or 'Deselect All' each time users enter the column. This is very confusing as thats not 'what' the checkbox within the tabel cell is about. Rather, it's the state of the overall selection. Instead, the table header should only communicate what the checkboxes are about.

Overall, 'Select All' or 'Deselect All' is content that is extraneous to the table semantics. While it may be convenient to visually place the 'Select / Deselect All' checkbox within this table header, that's less than ideal from a logical, semantic, and accessibility perspective.

This column header should just be named 'Selection'.

Instead, the 'Select / Deselect All'  checkbox should be moved out from the table, as that's content that doesn't semantically belong to the table.

As an example, I would like to suggest to have a look at what GMail does. Note: although the markup of the GMail list of messages is a HTML table, it's actually an ARIA grid. Regardless, some design concepts are applicable also to the WordPress tables.

![gmail](https://github.com/WordPress/gutenberg/assets/1682452/4256b2a4-2544-422c-bbe3-8f7771fa68ee)

It's not a case that in GMail, the 'Select / Deselect All' control is outside of the table / grid markup. Although GMail doesn't use table headers at all (as said that's actually an ARIA grid), the Google folks determined that 'Select / Deselect All'  doesn't belong to the table / grid semantic.

Similarly, I'd like to propose to move the data views 'Select / Deselect All' outside from the table.
- This would provide a more clear information to users an dsolve a long-standing problem with this control that comes from the classic implementation in core.
- It would make the table content way more understandable for screen reader users.
- In the future, it would provide a place whdre a more granular selection mechanism can be provided to users, like GMail dos, allowing selection of items based on a specific state or property: e.g. select all drafts, select all published, select all pending, etc.

### Step-by-step reproduction instructions

- Go to Site Editor > Pages > Manage all pages.
- Inspect the 'Select all' table header.
- Observe the only textual content of the table header is 'Select All' or 'Deselect All', depending on the selection state.

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes
